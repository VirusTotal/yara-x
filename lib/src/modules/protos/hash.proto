syntax = "proto2";
import "yara.proto";

package hash;

option (yara.module_options) = {
  name : "hash"
  root_message: "hash.Hash"
  rust_module: "hash"
  cargo_feature: "hash-module"
};

message Hash {
  // This module contains only exported functions, and doesn't return any data
}