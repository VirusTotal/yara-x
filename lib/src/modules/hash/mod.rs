use std::cell::RefCell;

use md5::Md5;
use rustc_hash::FxHashMap;
use sha1::Sha1;
use sha2::{Digest, Sha256};

use crate::modules::prelude::*;
use crate::modules::protos::hash::*;

#[cfg(test)]
mod tests;

thread_local!(
    static SHA256_CACHE: RefCell<FxHashMap<(i64, i64), String>> =
        RefCell::new(FxHashMap::default());

    static SHA1_CACHE: RefCell<FxHashMap<(i64, i64), String>> =
        RefCell::new(FxHashMap::default());

    static MD5_CACHE: RefCell<FxHashMap<(i64, i64), String>> =
        RefCell::new(FxHashMap::default());

    static CRC32_CACHE: RefCell<FxHashMap<(i64, i64), i64>> =
        RefCell::new(FxHashMap::default());

    static CHECKSUM32_CACHE: RefCell<FxHashMap<(i64, i64), i64>> =
        RefCell::new(FxHashMap::default());
);

#[module_main]
fn main(_data: &[u8]) -> Hash {
    // With every scanned file the cache must be cleared.
    SHA256_CACHE.with(|cache| cache.borrow_mut().clear());
    SHA1_CACHE.with(|cache| cache.borrow_mut().clear());
    MD5_CACHE.with(|cache| cache.borrow_mut().clear());
    CRC32_CACHE.with(|cache| cache.borrow_mut().clear());
    CHECKSUM32_CACHE.with(|cache| cache.borrow_mut().clear());

    Hash::new()
}

#[module_export(name = "md5")]
fn md5_data(
    ctx: &mut ScanContext,
    offset: i64,
    size: i64,
) -> Option<RuntimeString> {
    let cached = MD5_CACHE.with(|cache| -> Option<RuntimeString> {
        Some(RuntimeString::from_slice(
            ctx,
            cache.borrow().get(&(offset, size))?.as_bytes(),
        ))
    });

    if cached.is_some() {
        return cached;
    }

    let range = offset.try_into().ok()?..(offset + size).try_into().ok()?;
    let data = ctx.scanned_data().get(range)?;
    let mut hasher = Md5::new();

    hasher.update(data);

    let digest = format!("{:x}", hasher.finalize());

    MD5_CACHE.with(|cache| {
        cache.borrow_mut().insert((offset, size), digest.clone());
    });

    Some(RuntimeString::new(digest))
}

#[module_export(name = "md5")]
fn md5_str(ctx: &mut ScanContext, s: RuntimeString) -> Option<RuntimeString> {
    let mut hasher = Md5::new();
    hasher.update(s.as_bstr(ctx));

    Some(RuntimeString::new(format!("{:x}", hasher.finalize())))
}

#[module_export(name = "sha1")]
fn sha1_data(
    ctx: &mut ScanContext,
    offset: i64,
    size: i64,
) -> Option<RuntimeString> {
    let cached = SHA1_CACHE.with(|cache| -> Option<RuntimeString> {
        Some(RuntimeString::from_slice(
            ctx,
            cache.borrow().get(&(offset, size))?.as_bytes(),
        ))
    });

    if cached.is_some() {
        return cached;
    }

    let range = offset.try_into().ok()?..(offset + size).try_into().ok()?;
    let data = ctx.scanned_data().get(range)?;
    let mut hasher = Sha1::new();

    hasher.update(data);

    let digest = format!("{:x}", hasher.finalize());

    SHA1_CACHE.with(|cache| {
        cache.borrow_mut().insert((offset, size), digest.clone());
    });

    Some(RuntimeString::new(digest))
}

#[module_export(name = "sha1")]
fn sha1_str(ctx: &mut ScanContext, s: RuntimeString) -> Option<RuntimeString> {
    let mut hasher = Sha1::new();
    hasher.update(s.as_bstr(ctx));

    Some(RuntimeString::new(format!("{:x}", hasher.finalize())))
}

#[module_export(name = "sha256")]
fn sha256_data(
    ctx: &mut ScanContext,
    offset: i64,
    size: i64,
) -> Option<RuntimeString> {
    let cached = SHA256_CACHE.with(|cache| -> Option<RuntimeString> {
        Some(RuntimeString::from_slice(
            ctx,
            cache.borrow().get(&(offset, size))?.as_bytes(),
        ))
    });

    if cached.is_some() {
        return cached;
    }

    let range = offset.try_into().ok()?..(offset + size).try_into().ok()?;
    let data = ctx.scanned_data().get(range)?;
    let mut hasher = Sha256::new();

    hasher.update(data);

    let digest = format!("{:x}", hasher.finalize());

    SHA256_CACHE.with(|cache| {
        cache.borrow_mut().insert((offset, size), digest.clone());
    });

    Some(RuntimeString::new(digest))
}

#[module_export(name = "sha256")]
fn sha256_str(
    ctx: &mut ScanContext,
    s: RuntimeString,
) -> Option<RuntimeString> {
    let mut hasher = Sha256::new();
    hasher.update(s.as_bstr(ctx));

    Some(RuntimeString::new(format!("{:x}", hasher.finalize())))
}

#[module_export(name = "crc32")]
fn crc_data(ctx: &ScanContext, offset: i64, size: i64) -> Option<i64> {
    let cached = CRC32_CACHE.with(|cache| -> Option<i64> {
        Some(*cache.borrow().get(&(offset, size))?)
    });

    if cached.is_some() {
        return cached;
    }

    let range = offset.try_into().ok()?..(offset + size).try_into().ok()?;
    let data = ctx.scanned_data().get(range)?;
    let crc = crc32fast::hash(data);

    CRC32_CACHE.with(|cache| {
        cache.borrow_mut().insert((offset, size), crc.into());
    });

    Some(crc.into())
}

#[module_export(name = "crc32")]
fn crc_str(ctx: &ScanContext, s: RuntimeString) -> Option<i64> {
    let crc = crc32fast::hash(s.as_bstr(ctx));
    Some(crc.into())
}

#[module_export(name = "checksum32")]
fn checksum_data(ctx: &ScanContext, offset: i64, size: i64) -> Option<i64> {
    let cached = CHECKSUM32_CACHE.with(|cache| -> Option<i64> {
        Some(*cache.borrow().get(&(offset, size))?)
    });

    if cached.is_some() {
        return cached;
    }

    let range = offset.try_into().ok()?..(offset + size).try_into().ok()?;
    let data = ctx.scanned_data().get(range)?;
    let mut checksum = 0_u32;

    for byte in data {
        checksum = checksum.wrapping_add(*byte as u32)
    }

    CHECKSUM32_CACHE.with(|cache| {
        cache.borrow_mut().insert((offset, size), checksum.into());
    });

    Some(checksum.into())
}

#[module_export(name = "checksum32")]
fn checksum_str(ctx: &ScanContext, s: RuntimeString) -> Option<i64> {
    let mut checksum = 0_u32;
    for byte in s.as_bstr(ctx).as_bytes() {
        checksum = checksum.wrapping_add(*byte as u32)
    }
    Some(checksum.into())
}
