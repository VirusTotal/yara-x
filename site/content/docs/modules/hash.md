---
title: "hash"
description: ""
summary: ""
date: 2023-09-07T16:13:18+02:00
lastmod: 2023-09-07T16:13:18+02:00
draft: false
menu:
  docs:
    parent: ""
    identifier: "hash-module"
weight: 325
toc: true
seo:
  title: "" # custom title (optional)
  description: "" # custom description (recommended)
  canonical: "" # custom canonical URL (optional)
  noindex: false # false (default) or true
---

The `hash` module allows you to calculate hashes (MD5, SHA1, SHA256) and
checksums from portions of your file and create signatures based on those
hashes.

-------

## Functions

{{< callout context="caution" title="Important">}}

Hashes returned by the functions below are always in lowercase.

{{< /callout >}}

### md5(offset, size)

Returns the MD5 hash for size bytes starting at offset. When scanning a running
process the offset argument should be a virtual address within the process
address space. The returned string is always in lowercase.

Example: `hash.md5(0, filesize) == "feba6c919e3797e7778e8f2e85fa033d"`

### md5(string)

Returns the MD5 hash for the given string.

Example: `hash.md5("dummy") == "275876e34cf609db118f3d84b799a790"`

### sha1(offset, size)

Returns the SHA1 hash for the size bytes starting at offset. When scanning a
running process the offset argument should be a virtual address within the
process address space. The returned string is always in lowercase.

### sha1(string)

Returns the SHA1 hash for the given string.

### sha256(offset, size)

Returns the SHA256 hash for the size bytes starting at offset. When scanning a
running process the offset argument should be a virtual address within the
process address space. The returned string is always in lowercase.

### sha256(string)

Returns the SHA256 hash for the given string.

### checksum32(offset, size)

Returns a 32-bit checksum for the size bytes starting at offset. The checksum is
just the sum of all the bytes (unsigned).

### checksum32(string)

Returns a 32-bit checksum for the given string. The checksum is just the sum of
all the bytes in the string (unsigned).

### crc32(offset, size)

Returns a crc32 checksum for the size bytes starting at offset.

### crc32(string)

Returns a crc32 checksum for the given string.