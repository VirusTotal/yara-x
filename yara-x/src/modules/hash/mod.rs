use std::cell::RefCell;

use md5 as md5_hash;
use rustc_hash::FxHashMap;
use sha1::Sha1;
use sha2::{Digest, Sha256};

use crate::modules::prelude::*;
use crate::modules::protos::hash::*;

#[cfg(test)]
mod tests;

thread_local!(
    static SHA256_CACHE: RefCell<FxHashMap<(i64, i64), String>> =
        RefCell::new(FxHashMap::default());

    static SHA1_CACHE: RefCell<FxHashMap<(i64, i64), String>> =
        RefCell::new(FxHashMap::default());

    static MD5_CACHE: RefCell<FxHashMap<(i64, i64), String>> =
        RefCell::new(FxHashMap::default());
);

#[module_main]
fn main(_ctx: &ScanContext) -> Hash {
    // With every scanned file the cache must be cleared.
    SHA256_CACHE.with(|cache| cache.borrow_mut().clear());
    SHA1_CACHE.with(|cache| cache.borrow_mut().clear());
    MD5_CACHE.with(|cache| cache.borrow_mut().clear());

    Hash::new()
}

#[module_export(name = "md5")]
fn md5_data(
    ctx: &mut ScanContext,
    offset: i64,
    size: i64,
) -> Option<RuntimeString> {
    let cached = MD5_CACHE.with(|cache| -> Option<RuntimeString> {
        Some(RuntimeString::from_bytes(
            ctx,
            cache.borrow().get(&(offset, size))?,
        ))
    });

    if cached.is_some() {
        return cached;
    }

    let range = offset.try_into().ok()?..(offset + size).try_into().ok()?;
    let data = ctx.scanned_data().get(range)?;
    let digest = format!("{:x}", md5_hash::compute(data));
    let result = RuntimeString::from_bytes(ctx, digest.as_bytes());

    MD5_CACHE.with(|cache| {
        cache.borrow_mut().insert((offset, size), digest);
    });

    Some(result)
}

#[module_export(name = "md5")]
fn md5_str(ctx: &mut ScanContext, s: RuntimeString) -> Option<RuntimeString> {
    Some(RuntimeString::from_bytes(
        ctx,
        format!("{:x}", md5_hash::compute(s.as_bstr(ctx))),
    ))
}

#[module_export(name = "sha1")]
fn sha1_data(
    ctx: &mut ScanContext,
    offset: i64,
    size: i64,
) -> Option<RuntimeString> {
    let cached = SHA1_CACHE.with(|cache| -> Option<RuntimeString> {
        Some(RuntimeString::from_bytes(
            ctx,
            cache.borrow().get(&(offset, size))?,
        ))
    });

    if cached.is_some() {
        return cached;
    }

    let range = offset.try_into().ok()?..(offset + size).try_into().ok()?;
    let data = ctx.scanned_data().get(range)?;
    let mut hasher = Sha1::new();

    hasher.update(data);

    let digest = format!("{:x}", hasher.finalize());
    let result = RuntimeString::from_bytes(ctx, digest.as_bytes());

    SHA1_CACHE.with(|cache| {
        cache.borrow_mut().insert((offset, size), digest);
    });

    Some(result)
}

#[module_export(name = "sha1")]
fn sha1_str(ctx: &mut ScanContext, s: RuntimeString) -> Option<RuntimeString> {
    let mut hasher = Sha1::new();
    hasher.update(s.as_bstr(ctx));

    Some(RuntimeString::from_bytes(ctx, format!("{:x}", hasher.finalize())))
}

#[module_export(name = "sha256")]
fn sha256_data(
    ctx: &mut ScanContext,
    offset: i64,
    size: i64,
) -> Option<RuntimeString> {
    let cached = SHA256_CACHE.with(|cache| -> Option<RuntimeString> {
        Some(RuntimeString::from_bytes(
            ctx,
            cache.borrow().get(&(offset, size))?,
        ))
    });

    if cached.is_some() {
        return cached;
    }

    let range = offset.try_into().ok()?..(offset + size).try_into().ok()?;
    let data = ctx.scanned_data().get(range)?;
    let mut hasher = Sha256::new();

    hasher.update(data);

    let digest = format!("{:x}", hasher.finalize());
    let result = RuntimeString::from_bytes(ctx, digest.as_bytes());

    SHA256_CACHE.with(|cache| {
        cache.borrow_mut().insert((offset, size), digest);
    });

    Some(result)
}

#[module_export(name = "sha256")]
fn sha256_str(
    ctx: &mut ScanContext,
    s: RuntimeString,
) -> Option<RuntimeString> {
    let mut hasher = Sha256::new();
    hasher.update(s.as_bstr(ctx));

    Some(RuntimeString::from_bytes(ctx, format!("{:x}", hasher.finalize())))
}
