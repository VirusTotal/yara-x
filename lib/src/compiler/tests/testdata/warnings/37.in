import "hash"

rule test_1 { condition: hash.md5(0, filesize) == "AAB" }


rule test_2 { condition: "AD" == hash.sha256(0,filesize) }
