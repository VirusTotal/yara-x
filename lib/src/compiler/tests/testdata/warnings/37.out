warning[unsatisfiable_expr]: unsatisfiable expression
 --> line:3:31
  |
3 | rule test_1 { condition: hash.md5(0, filesize) == "AAB" }
  |                               ----------------    ----- the length of this string is 3
  |                               |
  |                               the length of this string is 32
  |
warning[unsatisfiable_expr]: unsatisfiable expression
 --> line:3:31
  |
3 | rule test_1 { condition: hash.md5(0, filesize) == "AAB" }
  |                               ----------------    ----- this contains uppercase characters
  |                               |
  |                               this is a lowercase string
  |
  = note: a lowercase string can't be equal to a string containing uppercase characters
warning[unsatisfiable_expr]: unsatisfiable expression
 --> line:6:39
  |
6 | rule test_2 { condition: "AD" == hash.sha256(0,filesize) }
  |                          ----         ------------------ the length of this string is 64
  |                          |
  |                          the length of this string is 2
  |
warning[unsatisfiable_expr]: unsatisfiable expression
 --> line:6:39
  |
6 | rule test_2 { condition: "AD" == hash.sha256(0,filesize) }
  |                          ----         ------------------ this is a lowercase string
  |                          |
  |                          this contains uppercase characters
  |
  = note: a lowercase string can't be equal to a string containing uppercase characters
